<a href="https://colab.research.google.com/github/MFaiqKhan/Sequential_MachineLearning/blob/main/LSTM_Word_Predictor_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# Building next word predictor using LSTM.

Dataset will be converted to input and output mapping and hence we will do supervised learning.
- Our Sentence have multiple Sentences and we will take it one by one then
by tokens though, like word by word and focus on the next word .
like :
"My name is faiq" .
it will be broken into input and output as:
input  output
- My  |  name
- My name |  is
- My name is | faiq

so like this, we will convert the whole data into input and output data .
- plus we have to it into numerical data .

In [ ]:
# Our Small Data :
data = """About the Program
What is the course fee for  Data Science Mentorship Program (DSMP 2023)
The course follows a monthly subscription model where you have to make monthly payments of Rs 799/month.
What is the total duration of the course?
The total duration of the course is 7 months. So the total course fee becomes 799*7 = Rs 5600(approx.)
What is the syllabus of the mentorship program?
We will be covering the following modules:
Python Fundamentals
Python libraries for Data Science
Data Analysis
SQL for Data Science
Maths for Machine Learning
ML Algorithms
Practical ML
MLOPs
Case studies
You can check the detailed syllabus here - https://learnwith.campusx.in/courses/CampusX-Data-Science-Mentorship-Program-637339afe4b0615a1bbed390
Will Deep Learning and NLP be a part of this program?
No, NLP and Deep Learning both are not a part of this program’s curriculum.
What if I miss a live session? Will I get a recording of the session?
Yes all our sessions are recorded, so even if you miss a session you can go back and watch the recording.
Where can I find the class schedule?
Checkout this google sheet to see month by month time table of the course - https://docs.google.com/spreadsheets/d/16OoTax_A6ORAeCg4emgexhqqPv3noQPYKU7RJ6ArOzk/edit?usp=sharing.
What is the time duration of all the live sessions?
Roughly, all the sessions last 2 hours.
What is the language spoken by the instructor during the sessions?
Hinglish
How will I be informed about the upcoming class?
You will get a mail from our side before every paid session once you become a paid user.
Can I do this course if I am from a non-tech background?
Yes, absolutely.
I am late, can I join the program in the middle?
Absolutely, you can join the program anytime.
If I join/pay in the middle, will I be able to see all the past lectures?
Yes, once you make the payment you will be able to see all the past content in your dashboard.
Where do I have to submit the task?
You don’t have to submit the task. We will provide you with the solutions, you have to self evaluate the task yourself.
Will we do case studies in the program?
Yes.
Where can we contact you?
You can mail us at nitish.campusx@gmail.com
Payment/Registration related questions
Where do we have to make our payments? Your YouTube channel or website?
You have to make all your monthly payments on our website. Here is the link for our website - https://learnwith.campusx.in/
Can we pay the entire amount of Rs 5600 all at once?
Unfortunately no, the program follows a monthly subscription model.
What is the validity of monthly subscription? Suppose if I pay on 15th Jan, then do I have to pay again on 1st Feb or 15th Feb
15th Feb. The validity period is 30 days from the day you make the payment. So essentially you can join anytime you don’t have to wait for a month to end.
What if I don’t like the course after making the payment. What is the refund policy?
You get a 7 days refund period from the day you have made the payment.
I am living outside India and I am not able to make the payment on the website, what should I do?
You have to contact us by sending a mail at nitish.campusx@gmail.com
Post registration queries
Till when can I view the paid videos on the website?
This one is tricky, so read carefully. You can watch the videos till your subscription is valid. Suppose you have purchased subscription on 21st Jan, you will be able to watch all the past paid sessions in the period of 21st Jan to 20th Feb. But after 21st Feb you will have to purchase the subscription again.
But once the course is over and you have paid us Rs 5600(or 7 installments of Rs 799) you will be able to watch the paid sessions till Aug 2024.
Why lifetime validity is not provided?
Because of the low course fee.
Where can I reach out in case of a doubt after the session?
You will have to fill a google form provided in your dashboard and our team will contact you for a 1 on 1 doubt clearance session
If I join the program late, can I still ask past week doubts?
Yes, just select past week doubt in the doubt clearance google form.
I am living outside India and I am not able to make the payment on the website, what should I do?
You have to contact us by sending a mail at nitish.campusx@gmai.com
Certificate and Placement Assistance related queries
What is the criteria to get the certificate?
There are 2 criterias:
You have to pay the entire fee of Rs 5600
You have to attempt all the course assessments.
I am joining late. How can I pay payment of the earlier months?
You will get a link to pay fee of earlier months in your dashboard once you pay for the current month.
I have read that Placement assistance is a part of this program. What comes under Placement assistance?
This is to clarify that Placement assistance does not mean Placement guarantee. So we dont guarantee you any jobs or for that matter even interview calls. So if you are planning to join this course just for placements, I am afraid you will be disappointed. Here is what comes under placement assistance
Portfolio Building sessions
Soft skill sessions
Sessions with industry mentors
Discussion on Job hunting strategies
"""

Assigning the index to all words in our data , through keras tokenizer class.

In [ ]:
import tensorflow as tf
from tensorflow.keras.preprocessing.text import Tokenizer

In [ ]:
tokenizer = Tokenizer()
tokenizer.fit_on_texts([data]) # we pass into list, as there can be multiple words obviously

In [ ]:
tokenizer.word_index

{'the': 1,
 'you': 2,
 'i': 3,
 'to': 4,
 'a': 5,
 'of': 6,
 'is': 7,
 'have': 8,
 'will': 9,
 'can': 10,
 'what': 11,
 'course': 12,
 'program': 13,
 'in': 14,
 'for': 15,
 'all': 16,
 'sessions': 17,
 'on': 18,
 'be': 19,
 'and': 20,
 'this': 21,
 'if': 22,
 'am': 23,
 'pay': 24,
 'payment': 25,
 'make': 26,
 'we': 27,
 'do': 28,
 'subscription': 29,
 'where': 30,
 'rs': 31,
 'so': 32,
 'campusx': 33,
 'session': 34,
 'our': 35,
 'paid': 36,
 'join': 37,
 'able': 38,
 'your': 39,
 'website': 40,
 'placement': 41,
 'fee': 42,
 'data': 43,
 'monthly': 44,
 'month': 45,
 'not': 46,
 'get': 47,
 'yes': 48,
 'once': 49,
 'past': 50,
 'feb': 51,
 'assistance': 52,
 'science': 53,
 '7': 54,
 '5600': 55,
 'are': 56,
 'watch': 57,
 'google': 58,
 'by': 59,
 'com': 60,
 'mail': 61,
 'from': 62,
 'contact': 63,
 'us': 64,
 'at': 65,
 'or': 66,
 'doubt': 67,
 'mentorship': 68,
 'payments': 69,
 '799': 70,
 'total': 71,
 'duration': 72,
 'months': 73,
 'learning': 74,
 'case': 75,
 'here': 76,
 '

Now we will go sentence by sentence and try to create this dataset :

My | name

My name | is

My name is | faiq

In [ ]:
# will break the sentence in our dataset by new line :

for sentence in data.split('\n'):
  print(sentence)

About the Program
What is the course fee for  Data Science Mentorship Program (DSMP 2023)
The course follows a monthly subscription model where you have to make monthly payments of Rs 799/month.
What is the total duration of the course?
The total duration of the course is 7 months. So the total course fee becomes 799*7 = Rs 5600(approx.)
What is the syllabus of the mentorship program?
We will be covering the following modules:
Python Fundamentals
Python libraries for Data Science
Data Analysis
SQL for Data Science
Maths for Machine Learning
ML Algorithms
Practical ML
MLOPs
Case studies
You can check the detailed syllabus here - https://learnwith.campusx.in/courses/CampusX-Data-Science-Mentorship-Program-637339afe4b0615a1bbed390
Will Deep Learning and NLP be a part of this program?
No, NLP and Deep Learning both are not a part of this program’s curriculum.
What if I miss a live session? Will I get a recording of the session?
Yes all our sessions are recorded, so even if you miss a sessi

In [ ]:
for sentence in data.split('\n'):
  print(tokenizer.texts_to_sequences([sentence])[0])

'''
data.split('\n'): This line splits the input data into individual sentences. The split function in Python splits a string into a list where each entry is a substring that was separated by the specified delimiter, in this case, the newline character ('\n').
for sentence in data.split('\n'): This line creates a loop that iterates over each sentence in the list returned by data.split('\n').
tokenizer.texts_to_sequences([sentence]): Inside the loop, the Keras Tokenizer's texts_to_sequences method is called on each sentence. This method transforms the sentence (which is a list of words) into a sequence of integers. It does this by looking up the integer that the Tokenizer has associated with each word during its fit_on_texts method. Unknown words (words that were not in the vocabulary when fit_on_texts was called) are skipped.
print(tokenizer.texts_to_sequences([sentence])): This line prints the sequence of integers that represents each sentence. Each unique word in the sentence is represented by a unique integer, according to the word-to-index mapping learned by the Tokenizer.
'''

[93, 1, 13]
[11, 7, 1, 12, 42, 15, 43, 53, 68, 13, 147, 148]
[1, 12, 94, 5, 44, 29, 95, 30, 2, 8, 4, 26, 44, 69, 6, 31, 70, 45]
[11, 7, 1, 71, 72, 6, 1, 12]
[1, 71, 72, 6, 1, 12, 7, 54, 73, 32, 1, 71, 12, 42, 149, 70, 54, 31, 55, 150]
[11, 7, 1, 96, 6, 1, 68, 13]
[27, 9, 19, 151, 1, 152, 153]
[97, 154]
[97, 155, 15, 43, 53]
[43, 156]
[157, 15, 43, 53]
[158, 15, 159, 74]
[98, 160]
[161, 98]
[162]
[75, 99]
[2, 10, 163, 1, 164, 96, 76, 77, 100, 33, 14, 165, 33, 43, 53, 68, 13, 166]
[9, 101, 74, 20, 102, 19, 5, 78, 6, 21, 13]
[103, 102, 20, 101, 74, 167, 56, 46, 5, 78, 6, 21, 168, 169]
[11, 22, 3, 104, 5, 105, 34, 9, 3, 47, 5, 106, 6, 1, 34]
[48, 16, 35, 17, 56, 170, 32, 107, 22, 2, 104, 5, 34, 2, 10, 171, 172, 20, 57, 1, 106]
[30, 10, 3, 173, 1, 108, 174]
[175, 21, 58, 176, 4, 79, 45, 59, 45, 109, 177, 6, 1, 12, 77, 178, 58, 60, 179, 180, 181, 182, 183, 184, 185]
[11, 7, 1, 109, 72, 6, 16, 1, 105, 17]
[186, 16, 1, 17, 187, 110, 188]
[11, 7, 1, 189, 190, 59, 1, 191, 192, 1, 17]
[193]
[111,

"\ndata.split('\n'): This line splits the input data into individual sentences. The split function in Python splits a string into a list where each entry is a substring that was separated by the specified delimiter, in this case, the newline character ('\n').\nfor sentence in data.split('\n'): This line creates a loop that iterates over each sentence in the list returned by data.split('\n').\ntokenizer.texts_to_sequences([sentence]): Inside the loop, the Keras Tokenizer's texts_to_sequences method is called on each sentence. This method transforms the sentence (which is a list of words) into a sequence of integers. It does this by looking up the integer that the Tokenizer has associated with each word during its fit_on_texts method. Unknown words (words that were not in the vocabulary when fit_on_texts was called) are skipped.\nprint(tokenizer.texts_to_sequences([sentence])): This line prints the sequence of integers that represents each sentence. Each unique word in the sentence is re

In [ ]:
input_sequences = []
for sentence in data.split('\n'):
  tokenized_sentence = tokenizer.texts_to_sequences([sentence])[0]

  for i in range (1, len(tokenized_sentence)):
    input_sequences.append(tokenized_sentence[:i+1])

for i in range (1, len(tokenized_sentence)): This nested loop iterates over the indices of the tokenized_sentence.
input_sentences.append(tokenized_sentence[:i+1]): This line appends the subsequence of tokenized_sentence from the start to the i+1th index to the input_sentences list. This is done for each index in the tokenized sentence, creating a list of all prefixes of the tokenized sentence.
In summary, this code is creating a list of all prefixes of each sentence in the input data, where each prefix and sentence is represented as a sequence of integers. This could be useful in a language model where the task is to predict the next word in a sentence, as each input sequence could be a prefix of a sentence and the corresponding output could be the next word

In [ ]:
input_sequences

[[93, 1],
 [93, 1, 13],
 [11, 7],
 [11, 7, 1],
 [11, 7, 1, 12],
 [11, 7, 1, 12, 42],
 [11, 7, 1, 12, 42, 15],
 [11, 7, 1, 12, 42, 15, 43],
 [11, 7, 1, 12, 42, 15, 43, 53],
 [11, 7, 1, 12, 42, 15, 43, 53, 68],
 [11, 7, 1, 12, 42, 15, 43, 53, 68, 13],
 [11, 7, 1, 12, 42, 15, 43, 53, 68, 13, 147],
 [11, 7, 1, 12, 42, 15, 43, 53, 68, 13, 147, 148],
 [1, 12],
 [1, 12, 94],
 [1, 12, 94, 5],
 [1, 12, 94, 5, 44],
 [1, 12, 94, 5, 44, 29],
 [1, 12, 94, 5, 44, 29, 95],
 [1, 12, 94, 5, 44, 29, 95, 30],
 [1, 12, 94, 5, 44, 29, 95, 30, 2],
 [1, 12, 94, 5, 44, 29, 95, 30, 2, 8],
 [1, 12, 94, 5, 44, 29, 95, 30, 2, 8, 4],
 [1, 12, 94, 5, 44, 29, 95, 30, 2, 8, 4, 26],
 [1, 12, 94, 5, 44, 29, 95, 30, 2, 8, 4, 26, 44],
 [1, 12, 94, 5, 44, 29, 95, 30, 2, 8, 4, 26, 44, 69],
 [1, 12, 94, 5, 44, 29, 95, 30, 2, 8, 4, 26, 44, 69, 6],
 [1, 12, 94, 5, 44, 29, 95, 30, 2, 8, 4, 26, 44, 69, 6, 31],
 [1, 12, 94, 5, 44, 29, 95, 30, 2, 8, 4, 26, 44, 69, 6, 31, 70],
 [1, 12, 94, 5, 44, 29, 95, 30, 2, 8, 4, 26, 44, 69, 6

In [ ]:
# will apply zero padding as the sizes of each numerical array is not same

max_len = max([len(x) for x in input_sequences])

In [ ]:
# Now we will do zero padding of max length to each sentence array
from tensorflow.keras.preprocessing.sequence import pad_sequences
padded_input_sequences = pad_sequences(input_sequences, maxlen = max_len, padding='pre')

This line uses the pad_sequences function from Keras to pad the sequences in input_sequences to have the same length. It adds zeroes to the beginning of each sequence to make each sequence the same length as the longest sequence[

In [ ]:
padded_input_sequences

array([[  0,   0,   0, ...,   0,  93,   1],
       [  0,   0,   0, ...,  93,   1,  13],
       [  0,   0,   0, ...,   0,  11,   7],
       ...,
       [  0,   0,   0, ..., 279,  18, 280],
       [  0,   0,   0, ...,  18, 280, 281],
       [  0,   0,   0, ..., 280, 281, 282]], dtype=int32)

In [ ]:
X = padded_input_sequences[:,:-1]
X # Our Input

array([[  0,   0,   0, ...,   0,   0,  93],
       [  0,   0,   0, ...,   0,  93,   1],
       [  0,   0,   0, ...,   0,   0,  11],
       ...,
       [  0,   0,   0, ...,   0, 279,  18],
       [  0,   0,   0, ..., 279,  18, 280],
       [  0,   0,   0, ...,  18, 280, 281]], dtype=int32)

In [ ]:
y = padded_input_sequences[:,-1]
y # Our Output

array([  1,  13,   7,   1,  12,  42,  15,  43,  53,  68,  13, 147, 148,
        12,  94,   5,  44,  29,  95,  30,   2,   8,   4,  26,  44,  69,
         6,  31,  70,  45,   7,   1,  71,  72,   6,   1,  12,  71,  72,
         6,   1,  12,   7,  54,  73,  32,   1,  71,  12,  42, 149,  70,
        54,  31,  55, 150,   7,   1,  96,   6,   1,  68,  13,   9,  19,
       151,   1, 152, 153, 154, 155,  15,  43,  53, 156,  15,  43,  53,
        15, 159,  74, 160,  98,  99,  10, 163,   1, 164,  96,  76,  77,
       100,  33,  14, 165,  33,  43,  53,  68,  13, 166, 101,  74,  20,
       102,  19,   5,  78,   6,  21,  13, 102,  20, 101,  74, 167,  56,
        46,   5,  78,   6,  21, 168, 169,  22,   3, 104,   5, 105,  34,
         9,   3,  47,   5, 106,   6,   1,  34,  16,  35,  17,  56, 170,
        32, 107,  22,   2, 104,   5,  34,   2,  10, 171, 172,  20,  57,
         1, 106,  10,   3, 173,   1, 108, 174,  21,  58, 176,   4,  79,
        45,  59,  45, 109, 177,   6,   1,  12,  77, 178,  58,  6

These lines split the padded sequences into inputs (X) and targets (y) for your model. The inputs to the model will be all but the last word of each sequence, and the target will be the last word of each sequence. This is consistent with a language model, where the task is to predict the next word in a sentence given the previous words

Padding is a preprocessing technique used in sequence prediction tasks such as natural language processing, time series forecasting, and speech processing. It is used to ensure that all sequences in a given batch have the same length, which is a requirement for most deep learning architectures[^9^].

In the context of your code snippet, padding is used to ensure that all sequences in `input_sequences` have the same length. This is done by adding zeroes to the beginning of each sequence until it has the same length as the longest sequence[^2^].

The reason why padding is necessary lies in the nature of neural networks. Many layers in neural networks, such as fully connected layers and convolutional layers, require fixed-size input. This is because these layers have a fixed number of weights, and each weight needs to be associated with an input value. If the size of the input varies, the network wouldn't know how to distribute its weights[^6^].

However, in sequence prediction tasks, it's common for individual samples to have different lengths. For example, in a language model, different sentences can have different numbers of words[^2^]. Padding is a way to handle this issue by making all sequences the same length, allowing them to be processed by the network.

While padding is a useful technique, it's important to note that it can introduce artificial data into your dataset. The zeroes that are added during padding don't carry any information and can potentially skew the results if not handled properly. This is why it's common to use a masking layer after padding to tell the network to ignore the padded parts of the input[^0^].

Sources:
- [Understanding Masking and Padding in Keras](https://tensorflow.rstudio.com/guides/keras/understanding_masking_and_padding)
- [Padding sequence data](https://www.tensorflow.org/guide/keras/understanding_masking_and_padding)
- [Why Is Padding Used?](https://www.baeldung.com/cs/deep-neural-networks-padding)
- [Introduction to Sequence Padding](https://saturncloud.io/blog/padding-sequences-of-2d-elements-in-keras-a-comprehensive-guide/)

The task can be either regression or classification depending on the nature of the output variable `y`.

If `y` represents a continuous output, then the task is a regression task. Regression predictive modeling is the task of approximating a mapping function from input variables to a continuous output variable[^1^]. An example of a regression task is predicting the price of a house based on features like its size and location[^1^].

If `y` represents discrete classes, then the task is a classification task. Classification predictive modeling is the task of approximating a mapping function from input variables to discrete categories[^1^]. An example of a classification task is predicting whether an email is spam or not based on features like the email's content and the sender's address[^1^].

Given the code snippet you provided, if the integer values in `y` represent different categories (for example, different words in a vocabulary), then this is a classification task. The model is trying to predict the next word in a sentence, which is a discrete class.

On the other hand, if the integer values in `y` represent a continuous quantity (for example, a count of something), then this is a regression task. However, this scenario is less likely in the context of your code snippet, as it's more common to predict the next word in a sentence (a classification task) rather than a continuous quantity.

Sources:
- [Difference Between Classification and Regression in Machine Learning](https://machinelearningmastery.com/classification-versus-regression-in-machine-learning/)
- [Classification vs. regression: What is the difference?](https://careerfoundry.com/en/blog/data-analytics/regression-vs-classification/)
- [Regression Versus Classification Machine Learning: What’s the Difference?](https://medium.com/quick-code/regression-versus-classification-machine-learning-whats-the-difference-345c56dd15f7)
- [Lasso Regression](https://geekflare.com/regression-vs-classification/)

In [ ]:
X.shape

(863, 56)

In [ ]:
y.shape

(863,)

In [ ]:
len(tokenizer.word_index)

282

In [ ]:
from tensorflow.keras.utils import to_categorical
y = to_categorical(y,num_classes=283)
# total words are our total classes btw, so it will make one hot encoding of total classes
# tokenizer output is starting from 1 index and one hot encoding always start from 0 , so I gave 283 instead of 282 so the last word can be processed too.

The code snippet you provided is converting the output variable `y` into a one-hot encoded format. This process is done in the context of a classification task.

One-hot encoding is a process of converting categorical data variables so they can be provided to machine learning algorithms to improve predictions. With one-hot, we convert each categorical value into a new categorical value and assign a binary value of 1 or 0. Each integer value is represented as a binary vector with all the values as zero except the index, which is marked with a 1[^0^].

In the context of Natural Language Processing (NLP) and your specific task, the `y` variable is likely representing the "next word" to predict in a sequence of words. Each unique word is represented by a unique integer. However, these integers imply an ordinal relationship among the words, which doesn't exist. To remove this implication, we use one-hot encoding. Each unique integer (word) is represented by a binary vector of length equal to the vocabulary size, with a 1 at the index corresponding to the word, and 0s elsewhere. This is a common practice in language models[^11^].

In Keras, the `to_categorical` function can be used to convert the `y` variable to one-hot encoded format:

```python
from tensorflow.keras.utils import to_categorical

y = to_categorical(y)
```

After this transformation, each entry in `y` is a one-hot encoded vector instead of an integer. The model then becomes a multi-class classification model where it predicts the probability of each word being the next word in the sentence[^0^].

In [ ]:
y.shape

(863, 283)

In [ ]:
y


array([[0., 1., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 1., 0., 0.],
       [0., 0., 0., ..., 0., 1., 0.],
       [0., 0., 0., ..., 0., 0., 1.]], dtype=float32)

In [ ]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, LSTM, Dense

In [ ]:
model = Sequential()
model.add(Embedding(283, 100, input_length=56))
model.add(LSTM(150))
model.add(Dense(283, activation='softmax'))

In [ ]:
model.compile(loss='categorical_crossentropy', optimizer='adam',metrics=['accuracy'])

The model you have is a Sequential model in Keras, which is a linear stack of layers that you can easily create by passing a list of layer instances to the constructor[^5^].

Let's break down each line of your model:

```python
model = Sequential()
```

This line initializes a new Sequential model[^5^].

```python
model.add(Embedding(283, 100, input_length=56))
```

This line adds an Embedding layer to the model. The Embedding layer is used to convert high-dimensional categorical data (in this case, words represented as integers) into a dense vector of fixed size[^9^]. This layer takes three arguments:

- `input_dim`: This is the size of the vocabulary, or the maximum integer index + 1. In your case, the vocabulary size is 283, which means you have 283 unique words in your text data[^9^].
- `output_dim`: This is the size of the dense vector. In your case, each word will be represented as a 100-dimensional vector[^9^].
- `input_length`: This is the length of input sequences. In your case, each input sequence has a length of 56[^9^].

```python
model.add(LSTM(150))
model.add(LSTM(150))
```

These lines add one LSTM (Long Short-Term Memory) layers to the model. LSTM is a type of Recurrent Neural Network (RNN) that can learn and remember over long sequences of data, making it highly effective for tasks like natural language processing[^1^]. Each LSTM layer in your model has 150 units, which is the dimensionality of the output space[^1^].

```python
model.add(Dense(283, activation='softmax'))
```

This line adds a Dense layer to the model. The Dense layer is a regular densely-connected neural network layer. It takes two arguments:

- `units`: This is the number of output neurons. In your case, the Dense layer has 283 neurons, which is equal to the size of your vocabulary[^9^].
- `activation`: This is the activation function to use. In your case, you're using the 'softmax' activation function, which is often used in the output layer of a network for multi-class classification problems[^9^].

In summary, your model begins with an Embedding layer that converts words into dense vectors. These vectors are then fed into two LSTM layers that learn features from the sequences of vectors. Finally, a Dense layer is used to output a probability distribution over the vocabulary for the next word in the sequence, making this a multi-class classification problem[^9^].

In [ ]:
model.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding_1 (Embedding)     (None, 56, 100)           28300     
                                                                 
 lstm_2 (LSTM)               (None, 150)               150600    
                                                                 
 dense (Dense)               (None, 283)               42733     
                                                                 
Total params: 221633 (865.75 KB)
Trainable params: 221633 (865.75 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


My apologies for the confusion. If there is only one LSTM layer in the model, the calculation would change accordingly.

Here's the revised breakdown:

```python
model.add(Embedding(283, 100, input_length=56))
```

The Embedding layer has `283 * 100 = 28,300` parameters. Each word in the vocabulary (283 words) is represented by a 100-dimensional vector, so there are `283 * 100` parameters[^14^].

```python
model.add(LSTM(150))
```

The LSTM layer has `(n*m + n^2 + n) * 4` parameters, where `n` is the number of units in the LSTM layer and `m` is the dimension of the input. In your case, the input to the LSTM layer is the output of the Embedding layer, which has 100 dimensions. So the number of parameters in the LSTM layer is `(150*100 + 150^2 + 150) * 4 = 150,600`[^12^].

```python
model.add(Dense(283, activation='softmax'))
```

The Dense layer has `(m * n + n)` parameters, where `m` is the dimension of the input and `n` is the number of units in the Dense layer. In your case, the input is the output of the LSTM layer, which has 150 dimensions. So the number of parameters in the Dense layer is `150 * 283 + 283 = 42,733`[^13^].

Adding up the parameters from all the layers, the total number of parameters in your model is `28,300 + 150,600 + 42,733 = 221,633`. This matches with the total parameters shown in your model summary.

The number of parameters in an LSTM layer can be calculated using the formula `(n*m + n^2 + n) * 4`, where `n` is the number of units in the LSTM layer and `m` is the dimension of the input .

Let's break this formula down:

1. `n*m`: This term represents the number of connections between the input and the LSTM units. Each LSTM unit is connected to every input, resulting in `n*m` connections .

2. `n^2`: This term represents the number of connections between the LSTM units themselves. Each LSTM unit is connected to every other LSTM unit, including itself, which results in `n^2` connections .

3. `n`: This term represents the bias units. Each LSTM unit has a bias, so there are `n` biases .

Finally, the entire sum `(n*m + n^2 + n)` is multiplied by 4. The reason for this is that an LSTM layer consists of four separate layers internally: the input gate, the forget gate, the memory cell, and the output gate. Each of these layers has `(n*m + n^2 + n)` parameters, so in total there are `4 * (n*m + n^2 + n)` parameters .


In [ ]:
model.fit(X,y,epochs=100)

Epoch 1/100
27/27 [==============================] - 7s 164ms/step - loss: 5.4577 - accuracy: 0.0707
Epoch 2/100
27/27 [==============================] - 3s 105ms/step - loss: 5.1023 - accuracy: 0.0776
Epoch 3/100
27/27 [==============================] - 3s 105ms/step - loss: 5.0267 - accuracy: 0.0776
Epoch 4/100
27/27 [==============================] - 3s 107ms/step - loss: 4.9906 - accuracy: 0.0776
Epoch 5/100
27/27 [==============================] - 4s 162ms/step - loss: 4.9495 - accuracy: 0.0776
Epoch 6/100
27/27 [==============================] - 3s 106ms/step - loss: 4.8799 - accuracy: 0.0776
Epoch 7/100
27/27 [==============================] - 3s 108ms/step - loss: 4.7623 - accuracy: 0.0892
Epoch 8/100
27/27 [==============================] - 3s 106ms/step - loss: 4.5798 - accuracy: 0.1217
Epoch 9/100
27/27 [==============================] - 4s 151ms/step - loss: 4.3634 - accuracy: 0.1564
Epoch 10/100
27/27 [==============================] - 3s 109ms/step - loss: 4.1149 - accura

In [ ]:
import numpy as np
# Predicting a single output next word
text = 'big'
# Text will be tokenized and padded then be put into model to get predicted the next word.

# tokenize
token_text = tokenizer.texts_to_sequences([text])[0]
# padding
padded_token_text = pad_sequences([token_text], maxlen = 56, padding='pre')
# predict
position = np.argmax(model.predict(padded_token_text))

1/1 [==============================] - 1s 701ms/step


In [ ]:
for word,index in tokenizer.word_index.items():
    if index == position:
      print(word)


nlp


In [ ]:
import time
# Predicting for Multiple wordsimport time
text = "I dont want"

for i in range(10):
  # tokenize
  token_text = tokenizer.texts_to_sequences([text])[0]
  # padding
  padded_token_text = pad_sequences([token_text], maxlen=56, padding='pre')
  # predict
  pos = np.argmax(model.predict(padded_token_text))

  for word,index in tokenizer.word_index.items():
    if index == pos:
      text = text + " " + word
      print(text)
      time.sleep(1)

1/1 [==============================] - 0s 37ms/step
I dont want living
1/1 [==============================] - 0s 28ms/step
I dont want living outside
1/1 [==============================] - 0s 25ms/step
I dont want living outside india
1/1 [==============================] - 0s 25ms/step
I dont want living outside india and
1/1 [==============================] - 0s 25ms/step
I dont want living outside india and i
1/1 [==============================] - 0s 25ms/step
I dont want living outside india and i am
1/1 [==============================] - 0s 29ms/step
I dont want living outside india and i am not
1/1 [==============================] - 0s 25ms/step
I dont want living outside india and i am not able
1/1 [==============================] - 0s 32ms/step
I dont want living outside india and i am not able to
1/1 [==============================] - 0s 25ms/step
I dont want living outside india and i am not able to make


To Improve more ACCURACY or make it work better or unseen data:
- we can train it on more data .
- hyperparameter tuning.
- optmizer change, lr change etc
-  can use advanced architecture like stacked LSTM, bidirectional LSTM, transformers etc.
